autonat.go

```diff
@@ -135,50 +135,78 @@ func (as *AmbientAutoNAT) autodetect() {
 	}
 
 	cli := NewAutoNATClient(as.host, as.getAddrs)
-	failures := 0
-
-	for _, pi := range peers {
-		ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-		as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, pstore.TempAddrTTL)
-		a, err := cli.DialBack(ctx, pi.ID)
-		cancel()
-
-		switch {
-		case err == nil:
-			log.Debugf("NAT status is public; address through %s: %s", pi.ID.Pretty(), a.String())
-			as.mx.Lock()
-			as.addr = a
-			as.status = NATStatusPublic
-			as.confidence = 0
-			as.mx.Unlock()
-			return
+	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
+	defer cancel()
+
+	var mx sync.Mutex
+	var pubaddr ma.Multiaddr
+	private := 0
+	public := 0
+
+	probe := 3 - as.confidence
+	if probe == 0 {
+		probe = 1
+	}
+	if probe > len(peers) {
+		probe = len(peers)
+	}
 
-		case IsDialError(err):
-			log.Debugf("dial error through %s: %s", pi.ID.Pretty(), err.Error())
-			failures++
-			if failures >= 3 || as.confidence >= 3 { // 3 times is enemy action
-				log.Debugf("NAT status is private")
-				as.mx.Lock()
-				as.status = NATStatusPrivate
-				as.confidence = 3
-				as.mx.Unlock()
-				return
+	var wg sync.WaitGroup
+
+	for _, pi := range peers[:probe] {
+		wg.Add(1)
+		go func(pi pstore.PeerInfo) {
+			as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, pstore.TempAddrTTL)
+			a, err := cli.DialBack(ctx, pi.ID)
+
+			switch {
+			case err == nil:
+				log.Debugf("Dialback through %s successful; public address is %s", pi.ID.Pretty(), a.String())
+				mx.Lock()
+				public++
+				pubaddr = a
```

Comment:
Is it possible that different peers will report different public endpoints, and we'll overwrite them here? This could happen if we're using different network interfaces for each peer, or if we're behind a corporate NAT that uses different exit points?

Reply:
we only report one address, so the override is inevitable. I don't think it matters much though, as we don't currently use the reported address for anything.